File: day9/day9.py
import matplotlib.pyplot as plt
import matplotlib.patches as patches
import time

start_time = time.time()
inp = list(map(lambda x: tuple(map(int, x.split(","))), open("9").read().splitlines()))


def area(p1, p2):
    return (abs(p1[0] - p2[0]) + 1) * (abs(p1[1] - p2[1]) + 1)


max_area = 0

for i in range(len(inp)):
    for j in range(i + 1, len(inp)):
        max_area = max(max_area, area(inp[i], inp[j]))

print(f"Part 1: {max_area}")

vertical_lines = sorted(
    [sorted(inp[i : i + 2], key=lambda x: x[1]) for i in range(0, len(inp), 2)],
    key=lambda x: x[0][0],
)

horizontal_lines = [
    sorted(inp[i : i + 2], key=lambda x: x[0]) for i in range(1, len(inp) - 1, 2)
]


def in_poly(p):
    # Ray cast to left
    num_crossings = 0
    lows = set()
    highs = set()
    for line in vertical_lines:
        if line[0][0] >= p[0]:
            break
        if line[0][1] <= p[1] <= line[1][1]:
            if line[0][1] == p[1] and p[1] in highs:
                continue
            if line[1][1] == p[1] and p[1] in lows:
                continue
            lows.add(line[0][1])
            highs.add(line[1][1])
            num_crossings += 1

    return num_crossings % 2 == 1


longest_two = sorted(
    horizontal_lines, key=lambda x: abs(x[0][0] - x[1][0]), reverse=True
)
mid1 = (
    longest_two[0][0]
    if longest_two[0][0][0] > longest_two[0][1][0]
    else longest_two[0][1]
)
mid2 = (
    longest_two[1][0]
    if longest_two[1][0][0] > longest_two[1][1][0]
    else longest_two[1][1]
)

if mid1[1] < mid2[1]:
    mid1, mid2 = mid2, mid1

mid1_ylimit = 0
mid2_ylimit = 0
for i in horizontal_lines:
    if i[0][1] > mid1[1] and i[0][0] <= mid1[0] <= i[1][0]:
        mid1_ylimit = i[0][1]
    if i[0][1] < mid2[1] and i[0][0] <= mid2[0] <= i[1][0]:
        mid2_ylimit = i[0][1]


point = []
max_area = 0

for p in inp:
    m = None
    if p[0] >= mid1[0]:
        continue
    if mid1[1] <= p[1] <= mid1_ylimit:
        m = mid1
    elif mid2[1] <= p[1] <= mid2_ylimit:
        m = mid2
    if m is not None:
        p1, p2 = (p[0], m[1]), (m[0], p[1])
        if (a := area(m, p)) > max_area and in_poly(p1) and in_poly(p2):
            max_area = a
            point = [p if p[1] < m[1] else p1, abs(p[0] - m[0]), abs(p[1] - m[1])]
print(f"Part 2: {max_area}")
print(f"Execution time: {time.time() - start_time}")
rect = patches.Rectangle(
    point[0],
    point[1],
    point[2],
    facecolor="lightblue",
    edgecolor="blue",
    linewidth=2,
)

fig = plt.figure()
ax = fig.add_subplot(111)
ax.add_patch(rect)
x = [i[0] for i in inp]
y = [i[1] for i in inp]
for i in range(len(inp) - 1):
    plt.plot(x[i : i + 2], y[i : i + 2], "ro-")
plt.show()

File: day9/day9.java
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashSet;
import java.util.List;
import java.util.Scanner;
import java.util.Set;

public class day9 {

    public static void main(String[] args) {
        List<Point> points = new ArrayList<>();
        Scanner sc = new Scanner(System.in);

        while (sc.hasNext()) {
            String[] line = sc.next().split(",");
            points.add(new Point(line));
        }

        sc.close();

        List<Line> verticals = new ArrayList<>();
        List<Line> horizontals = new ArrayList<>();
        long part1 = 0;

        for (int i = 0; i < points.size(); i++) {
            for (int j = i + 1; j < points.size(); j++) {
                part1 = Math.max(part1, Point.area(points.get(i), points.get(j)));
            }

            if (i % 2 == 0)
                verticals.add(new Line(points.get(i), points.get(i + 1)));
            else if (i + 1 < points.size())
                horizontals.add(new Line(points.get(i), points.get(i + 1)));
        }

        System.out.println(String.format("Part 1: %d", part1));

        verticals.sort(new Comparator<Line>() {
            @Override
            public int compare(Line arg0, Line arg1) {
                return arg0.p1.x - arg1.p1.x;
            }
        });
        verticals.sort(new Comparator<Line>() {
            @Override
            public int compare(Line arg0, Line arg1) {
                return arg0.p1.x - arg1.p1.x;
            }
        });

        horizontals.sort(new Comparator<Line>() {
            @Override
            public int compare(Line arg0, Line arg1) {
                return arg1.horizontalLength() - arg0.horizontalLength();
            }
        });

        Point mid1 = horizontals.get(0).getRightPoint();
        Point mid2 = horizontals.get(1).getRightPoint();

        if (mid1.y < mid2.y) {
            Point temp = mid2;
            mid2 = mid1;
            mid1 = temp;
        }

        int mid1_ylimit = 0;
        int mid2_ylimit = 0;

        for (Line line : horizontals) {
            int left = line.getLeft();
            int right = line.getRight();
            if (line.p1.y > mid1.y && left <= mid1.x && mid1.x <= right)
                mid1_ylimit = line.p1.y;
            if (line.p1.y < mid2.y && left <= mid2.x && mid2.x <= right)
                mid2_ylimit = line.p1.y;
        }

        long part2 = 0;

        for (Point point : points) {
            Point m = null;

            if (point.x >= mid1.x)
                continue;

            if (mid1.y <= point.y && point.y <= mid1_ylimit)
                m = mid1;
            else if (mid2.y <= point.y && point.y <= mid2_ylimit)
                m = mid2;

            if (m != null) {
                Point p1 = new Point(point.x, m.y);
                Point p2 = new Point(m.x, point.y);

                if (Point.area(m, point) > part2 && p1.inPolygon(verticals) && p2.inPolygon(verticals))
                    part2 = Point.area(m, point);
            }
        }

        System.out.println(String.format("Part 2: %d", part2));
    }
}

class Point {
    int x;
    int y;

    Point(int x, int y) {
        this.x = x;
        this.y = y;
    }

    Point(String[] line) {
        this.x = Integer.valueOf(line[0]);
        this.y = Integer.valueOf(line[1]);
    }

    static long area(Point p1, Point p2) {
        return Math.multiplyFull((Math.abs(p1.x - p2.x) + 1), (Math.abs(p1.y - p2.y) + 1));
    }

    // Assumes verticals are sorted.
    boolean inPolygon(List<Line> verticals) {
        int num_crossings = 0;

        Set<Integer> lows = new HashSet<>();
        Set<Integer> highs = new HashSet<>();

        for (Line line : verticals) {
            if (line.p1.x >= this.x)
                break;

            int lower = line.getLower();
            int higher = line.getHigher();
            if (lower <= this.y && this.y <= higher) {
                if (lower == this.y && highs.contains(this.y))
                    continue;
                if (higher == this.y && lows.contains(this.y))
                    continue;

                lows.add(lower);
                highs.add(higher);
                num_crossings += 1;
            }
        }

        return num_crossings % 2 == 1;
    }

    @Override
    public String toString() {
        return String.format("(%s, %s)", x, y);
    }
}

class Line {
    Point p1;
    Point p2;

    Line(Point p1, Point p2) {
        this.p1 = p1;
        this.p2 = p2;
    }

    int getLeft() {
        return Math.min(p1.x, p2.x);
    }

    int getRight() {
        return Math.max(p1.x, p2.x);
    }

    int getHigher() {
        return Math.max(p1.y, p2.y);
    }

    int getLower() {
        return Math.min(p1.y, p2.y);
    }

    Point getRightPoint() {
        return p1.x > p2.x ? p1 : p2;
    }

    int horizontalLength() {
        return Math.abs(p1.x - p2.x);
    }

    @Override
    public String toString() {
        return String.format("[%s, %s]", p1, p2);
    }
}
